### ISSUE TITLE
Patch Application Relies on Global State and `AutoLigerKernelForCausalLM.from_config` doesn't work properly

### ISSUE THREAD

**Issue by lapp0** [linkedin/Liger-Kernel#242]
### 🐛 Describe the bug

I cannot apply `AutoLigerKernelForCausalLM.from_config` unless I first call `AutoLigerKernelForCausalLM.from_pretrained`.

The side effect of being patchable after `from_pretrained` is called is because liger patches modules, rather than objects. This side effect is separately problematic because `transformers.AutoModelForCausalLM` will produce liger-kernel models as well once the patch is applied.

Ideally
- 1) Liger patches objects not modules (this is especially important to me because I **only** apply **forward pass** to one of the models in my distillation trainer)
- 2) `from_config` works without first calling `from_pretrained`

## Doesn't Patch

```python3
from liger_kernel.transformers.auto_model import AutoLigerKernelForCausalLM
import transformers

model = AutoLigerKernelForCausalLM.from_config(
    transformers.AutoConfig.from_pretrained("HuggingFaceTB/SmolLM-135M")
)
print(model)
```

Output:
```
LlamaForCausalLM(
  (model): LlamaModel(
    (embed_tokens): Embedding(49152, 576)
    (layers): ModuleList(
      (0-29): 30 x LlamaDecoderLayer(
        (self_attn): LlamaSdpaAttention(
          (q_proj): Linear(in_features=576, out_features=576, bias=False)
          (k_proj): Linear(in_features=576, out_features=192, bias=False)
          (v_proj): Linear(in_features=576, out_features=192, bias=False)
          (o_proj): Linear(in_features=576, out_features=576, bias=False)
          (rotary_emb): LlamaRotaryEmbedding()
        )
        (mlp): LlamaMLP(
          (gate_proj): Linear(in_features=576, out_features=1536, bias=False)
          (up_proj): Linear(in_features=576, out_features=1536, bias=False)
          (down_proj): Linear(in_features=1536, out_features=576, bias=False)
          (act_fn): SiLU()
        )
        (input_layernorm): LlamaRMSNorm((576,), eps=1e-05)
        (post_attention_layernorm): LlamaRMSNorm((576,), eps=1e-05)
      )
    )
    (norm): LlamaRMSNorm((576,), eps=1e-05)
    (rotary_emb): LlamaRotaryEmbedding()
  )
  (lm_head): Linear(in_features=576, out_features=49152, bias=False)
)
```

## This hack works

```python3
from liger_kernel.transformers.auto_model import AutoLigerKernelForCausalLM
import transformers

# HACK: this model is never used
AutoLigerKernelForCausalLM.from_pretrained("HuggingFaceTB/SmolLM-135M")

model = AutoLigerKernelForCausalLM.from_config(
    transformers.AutoConfig.from_pretrained("HuggingFaceTB/SmolLM-135M")
)
print(model)
```

Output
```
LlamaForCausalLM(
  (model): LlamaModel(
    (embed_tokens): Embedding(49152, 576)
    (layers): ModuleList(
      (0-29): 30 x LlamaDecoderLayer(
        (self_attn): LlamaSdpaAttention(
          (q_proj): Linear(in_features=576, out_features=576, bias=False)
          (k_proj): Linear(in_features=576, out_features=192, bias=False)
          (v_proj): Linear(in_features=576, out_features=192, bias=False)
          (o_proj): Linear(in_features=576, out_features=576, bias=False)
          (rotary_emb): LlamaRotaryEmbedding()
        )
        (mlp): LigerSwiGLUMLP(
          (gate_proj): Linear(in_features=576, out_features=1536, bias=False)
          (up_proj): Linear(in_features=576, out_features=1536, bias=False)
          (down_proj): Linear(in_features=1536, out_features=576, bias=False)
        )
        (input_layernorm): LigerRMSNorm((576,), eps=1e-05, offset=0.0)
        (post_attention_layernorm): LigerRMSNorm((576,), eps=1e-05, offset=0.0)
      )
    )
    (norm): LigerRMSNorm((576,), eps=1e-05, offset=0.0)
    (rotary_emb): LlamaRotaryEmbedding()
  )
  (lm_head): Linear(in_features=576, out_features=49152, bias=False)
)
```

### Versions

Environment Report:
-------------------
Operating System: Linux-5.15.0-117-generic-x86_64-with-glibc2.35
Python version: 3.10.14
PyTorch version: 2.4.1+cu121
CUDA version: 12.1
Triton version: 3.0.0
Transformers version: 4.44.2

**Comment by lapp0** [comment]
Another issue with `from_config`, even with this hack. I need to explicitly remove the `labels` argument from the tokenizer to allow the forward pass to return logits.

```python3
>>> import transformers
>>> from liger_kernel.transformers.auto_model import AutoLigerKernelForCausalLM
>>> import torch

>>> from liger_kernel.transformers.auto_model import AutoLigerKernelForCausalLM

# HACK
>>> AutoLigerKernelForCausalLM.from_pretrained("HuggingFaceTB/SmolLM-135M", device_map="cuda").to("cuda")

# Real model creation
>>> model_liger = AutoLigerKernelForCausalLM.from_config(transformers.AutoConfig.from_pretrained("HuggingFaceTB/SmolLM-135M")).to("cuda")

>>> input_ids = torch.tensor([[1,2],[3,4]]).to("cuda")

# GOOD
>>> model_liger(torch.tensor([[1,2],[3,4]]).to("cuda")).logits                                           
tensor([[[-0.5535,  0.5003, -0.3423,  ..., -0.2395, -0.8354,  0.5833],                                   
         [-0.5019,  0.1111, -0.1496,  ..., -0.1092, -0.8418,  0.4779]],                                                                                                                                           
                                                                                                                                                                                                                  
        [[-0.0211, -0.0767, -0.7870,  ...,  0.6154,  0.0326,  0.7117],                                                                                                                                            
         [-0.2163,  0.0827, -0.6699,  ...,  0.6046,  0.2357,  0.6225]]],                                 
       device='cuda:0', grad_fn=<UnsafeViewBackward0>)                                                   

# BAD
>>> model_liger(input_ids=input_ids, labels=input_ids).logits
>>> 
```

**Comment by yundai424** [comment]
hi @lapp0 just to understand, [this](https://github.com/linkedin/Liger-Kernel/blob/main/README.md#2-apply-model-specific-patching-apis) API doesn't apply to you because you don't want the other models to be patched (also would like to understand why we can't do it on both teacher and student models)?

1. For the `AutoLigerKernelForCausalLM.from_config` issue, one quick workaround before it getting fixed is to use [another API](https://github.com/linkedin/Liger-Kernel/blob/main/README.md#2-apply-model-specific-patching-apis) before instantiating the model. This doesn't meet your need of only applying to forward to one model (i'm assuming you want it applied to teacher cuz it doesn't do backward) though. Hacky workaround is to follow these steps: 1) instantiate student model, 2) `apply_liger_kernel_to_llama`, 3) instantiate your teacher model
2. for object patching, i think we can go with two approaches: 
    - provide a context manager, only in which new model creation will use the patch, similar to DeepSpeed zero3 https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/partition_parameters.py#L320-L344
    - pure object patching (potentially leveraging `MethodType`)
@shimizust @ByronHsu wdyt?

**Comment by lapp0** [comment]
>also would like to understand why we can't do it on both teacher and student models

Is Liger optimized for improvements in forward pass, or only backwards pass? It's unclear to me which component the benchmarks apply to what. I'm seeing basically no improvement when I apply liger kernel to the teacher model, but ~10% train throughput increase when I apply it to the student model. 

Perhaps the issue is with `use_liger_kernel` in huggingface training arguments, I'd have to investigate further.

**Comment by yundai424** [comment]
Both forward and backward for all kernels. But yeah it'll have more significant impact on models that will perform backward - not only just trim the computation overhead but also avoid storing all the intermediate activation step-to-step out of each op in native pytorch implementation. 

It'll be interesting to take a look at trace view in profiling result with & w/o applying to teacher to tell what could be the reason of not seeing improvement 🤔

**Comment by shimizust** [comment]
@lapp0 I think this PR will enable patching post model initialization: https://github.com/linkedin/Liger-Kernel/pull/199

You would pass your model into: `_apply_liger_kernel_to_instance(model=model)`.

**Comment by lapp0** [comment]
This is exactly what I need @shimizust , thanks so much for your awesome work, along with everyone else in this thread and on the team!

**Comment by ByronHsu** [comment]
PR is merged!
